src/realm/table.cpp

```diff
@@ -2535,24 +2535,44 @@ MemStats Table::stats() const
 }
 #endif // LCOV_EXCL_STOP ignore debug functions
 
-Obj Table::create_object(ObjKey key, const FieldValues& values)
+Obj Table::create_object_internal(ObjKey key, const FieldValues& values)
 {
+    // Create an object after detecting tombstones, hash collisions, etc.
+
     if (m_is_embedded)
         throw LogicError(LogicError::wrong_kind_of_table);
-    if (key == null_key) {
-        GlobalKey object_id = allocate_object_id_squeezed();
-        key = object_id.get_local_key(get_sync_file_id());
-        if (auto repl = get_repl())
-            repl->create_object(this, object_id);
-    }
 
+    // Tombstones and null-keys not allowed.
     REALM_ASSERT(key.value >= 0);
 
     bump_content_version();
     bump_storage_version();
-    Obj obj = m_clusters.insert(key, values);
+    return m_clusters.insert(key, values);
+}
 
-    return obj;
+Obj Table::create_object(ObjKey key, const FieldValues& values)
+{
+    if (m_is_embedded)
+        throw LogicError(LogicError::wrong_kind_of_table);
+    REALM_ASSERT(!get_primary_key_column());
+
+    GlobalKey object_id;
+    if (key == null_key) {
+        object_id = allocate_object_id_squeezed();
+        key = object_id.get_local_key(get_sync_file_id());
+    }
+    else {
+        // Tombstones not allowed.
+        REALM_ASSERT(key.value >= 0);
+        if (is_valid(key)) {
+            throw InvalidKey{"requested object key already in use"};
+        }
+        object_id = get_object_id(key);
+    }
+
+    // Forward to the GlobalKey version of create_object(), so it can revive
+    // tombstones if needed.
```

Comment:
When would there be tombstones to revive here? Wouldn't that only happen if allocate_object_id_squeezed() produced a key that it can locally determine is already in use? It seems like allocate_object_id_squeezed() should just not do that.

Reply:
Good catch, it looks like the line above is wrong. It should call `GlobalKey{ObjKey, sync_file_id}`.